bootstore/src/schemes/v0/peer_networking.rs

```diff
@@ -0,0 +1,635 @@
+            write_queue: VecDeque::with_capacity(MSG_WRITE_QUEUE_CAPACITY),
+            current_write: Cursor::new(Vec::new()),
+        }
+    }
+
+    // Run the main loop of the connection
+    //
+    // The task can only return a `ConnErr`, otherwise it runs forever
+    async fn run(&mut self) -> ConnErr {
+        let mut interval = interval(PING_INTERVAL);
+        interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+
+        loop {
+            if !self.current_write.has_remaining()
+                && !self.write_queue.is_empty()
+            {
+                self.current_write =
+                    Cursor::new(self.write_queue.pop_front().unwrap());
+            }
+
+            let res = tokio::select! {
+                _ = interval.tick() => {
+                    self.ping().await
+                }
+                Some(msg) = self.rx.recv() => {
+                    self.on_msg_from_main(msg).await
+                }
+                res = self.read_sock.read(&mut self.read_buf[self.total_read..]) => {
+                    self.read(res).await
+                }
+                res = self.write_sock.write_buf(&mut self.current_write),
+                   if self.current_write.has_remaining() => {
+                        self.check_write_result(res).await
+                }
+            };
+
+            if let Err(err) = res {
+                warn!(self.log, "Connection error: {:?}", err);
+                return err;
+            }
+        }
+    }
+
+    async fn on_msg_from_main(
+        &mut self,
+        msg: MainToConnMsg,
+    ) -> Result<(), ConnErr> {
+        match msg {
+            MainToConnMsg::Close => {
+                let _ = self.close().await;
+                return Err(ConnErr::Close);
+            }
+            MainToConnMsg::Msg(msg) => {
+                if self.write_queue.len() == MSG_WRITE_QUEUE_CAPACITY {
+                    warn!(self.log, "Closing connection: write queue full");
+                    self.close().await
+                } else {
+                    match write_framed(&msg) {
+                        Ok(msg) => {
+                            self.write_queue.push_back(msg);
+                            Ok(())
+                        }
+                        Err(e) => {
+                            warn!(
+                                self.log,
+                                "Closing connection: Failed to serialize msg: {}",
+                                e
+                            );
+                            self.close().await
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    async fn check_write_result(
+        &mut self,
+        res: Result<usize, std::io::Error>,
+    ) -> Result<(), ConnErr> {
+        match res {
+            Ok(_) => {
+                if !self.current_write.has_remaining() {
+                    self.current_write = Cursor::new(Vec::new());
```

Comment:
```suggestion
                    self.current_write = Cursor::default();
```

Reply:
(also, this suggests that `current_write` should be an `Option<Cursor<Vec<u8>>>`, instead of using an empty cursor to indicate absence)

Reply:
Yes. I actually had an `Option<Cursor<Vec<u8>>>` at first, but while iterating I changed it due to a limitation in one version of the code. I don't recall what that was, but I also don't think its relevant any more. I will change it back.

Reply:
I think `Option<T>` might be a bit more annoying to use -- `has_remaining` is an easier check.

Reply:
In the interest of urgency, I'm also gonna skip this for now.